src/neps/optimizers/bayesian_optimization/acquisition_functions/mf_ei.py

```diff
@@ -2,41 +2,98 @@
 
     def get_budget_level(self, config) -> int:
         return int((config.fidelity.value - config.fidelity.lower) / self.b_step)
 
-    def preprocess(self, x: Iterable) -> Tuple[Iterable, Iterable]:
+    # def _preprocess_tabular(self, x: pd.Series) -> pd.Series:
+    #     if len(x) == 0:
+    #         return x
+    #     # extract fid name
+    #     _x = x.loc[0].hp_values()
+    #     _x.pop("id")
+    #     fid_name = list(_x.keys())[0]
+    #     for i in x.index.values:
+    #         # extracting actual HPs from the tabular space
+    #         _config = self.pipeline_space.custom_grid_table.loc[x.loc[i]["id"].value].to_dict()
+    #         # updating fidelities as per the candidate set passed
+    #         _config.update({fid_name: x.loc[i][fid_name].value})
+    #         # placeholder config from the raw tabular space
+    #         config = self.pipeline_space.raw_tabular_space.sample(
+    #             patience=100, 
+    #             user_priors=True, 
+    #             ignore_fidelity=True  # True allows fidelity to appear in the sample
+    #         )
+    #         # copying values from table to placeholder config of type SearchSpace
+    #         config.load_from(_config)
+    #         # replacing the ID in the candidate set with the actual HPs of the config
+    #         x.loc[i] = config
+    #     return x
+
+    def preprocess(self, x: pd.Series) -> Tuple[Iterable, Iterable]:
         """Prepares the configurations for appropriate EI calculation.
 
         Takes a set of points and computes the budget and incumbent for each point, as
         required by the multi-fidelity Expected Improvement acquisition function.
         """
         budget_list = []
-        performances = self.observations.get_best_performance_for_each_budget()
 
-        for _x in x:
-            budget_list.append(self.get_budget_level(_x))
+        if self.pipeline_space.has_tabular:
+            # preprocess tabular space differently
+            # expected input: IDs pertaining to the tabular data
+            # expected output: IDs pertaining to current observations and set of HPs
+            # x = self._preprocess_tabular(x)
+            x = map_real_hyperparameters_from_tabular_ids(x, self.pipeline_space)
+        indices_to_drop = []
+        for i, config in x.items():
+            target_fidelity = config.fidelity.lower
+            if i <= max(self.observations.seen_config_ids):
+                # IMPORTANT to set the fidelity at which EI will be calculated only for
+                # the partial configs that have been observed already
+                target_fidelity = config.fidelity.value + self.b_step
+                config.fidelity.value = min(
+                    target_fidelity, config.fidelity.upper
+                )  # to respect the bounded fidelity
+            else:
+                config.fidelity.value = target_fidelity
+
+            if np.isclose(target_fidelity, config.fidelity.value):
```

Comment:
Moving this check into the above `if` statement will reduce the `numpy` comparisons when the `SAMPLES_TO_DRAW` is high

Reply:
says `outdated` now and the code seems to have changed...